From 00bb043568dd326e73f550dd6f4a4f211f771884 Mon Sep 17 00:00:00 2001
From: Igor Opaniuk <igor.opaniuk@foundries.io>
Date: Mon, 19 Apr 2021 23:51:42 +0300
Subject: [PATCH 140/322] [FIO internal] mx6: add function for checking board
 close state

Introduce function for reading SEC_CONFIG[1] state on MX6 SoCs, which
can be used to change boot logic based on board open/close state.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>
---
 arch/arm/include/asm/arch-mx6/imx-regs.h  | 1 +
 arch/arm/include/asm/arch-mx6/sys_proto.h | 1 +
 arch/arm/mach-imx/mx6/soc.c               | 7 +++++++
 3 files changed, 9 insertions(+)

diff --git a/arch/arm/include/asm/arch-mx6/imx-regs.h b/arch/arm/include/asm/arch-mx6/imx-regs.h
index 63746da2548..6df5b5b580c 100644
--- a/arch/arm/include/asm/arch-mx6/imx-regs.h
+++ b/arch/arm/include/asm/arch-mx6/imx-regs.h
@@ -457,6 +457,7 @@ extern void imx_get_mac_from_fuse(int dev_id, unsigned char *mac);
 #define SRC_SCR_CORE_3_ENABLE_MASK      (1<<SRC_SCR_CORE_3_ENABLE_OFFSET)
 
 #define SRC_GPR10_PERSIST_SECONDARY_BOOT	BIT(30)
+#define SRC_SBMR2_SEC_CONFIG			BIT(1)
 
 struct rdc_regs {
 	u32	vir;		/* Version information */
diff --git a/arch/arm/include/asm/arch-mx6/sys_proto.h b/arch/arm/include/asm/arch-mx6/sys_proto.h
index 9a29e66074a..af197d2f44e 100644
--- a/arch/arm/include/asm/arch-mx6/sys_proto.h
+++ b/arch/arm/include/asm/arch-mx6/sys_proto.h
@@ -52,5 +52,6 @@ void finish_anatop_bypass(void);
 
 int boot_mode_getprisec(void);
 void boot_mode_enable_secondary(bool enable);
+int boot_mode_is_closed(void);
 
 #endif /* __SYS_PROTO_IMX6_ */
diff --git a/arch/arm/mach-imx/mx6/soc.c b/arch/arm/mach-imx/mx6/soc.c
index b00552bdeff..3656c804437 100644
--- a/arch/arm/mach-imx/mx6/soc.c
+++ b/arch/arm/mach-imx/mx6/soc.c
@@ -689,6 +689,13 @@ void boot_mode_enable_secondary(bool enable)
 		clrbits_le32(&psrc->gpr10, persist_sec);
 }
 
+int boot_mode_is_closed(void)
+{
+	struct src *psrc = (struct src *)SRC_BASE_ADDR;
+
+	return !!(readl(&psrc->sbmr2) & SRC_SBMR2_SEC_CONFIG);
+}
+
 int boot_mode_getprisec(void)
 {
 	struct src *psrc = (struct src *)SRC_BASE_ADDR;
-- 
2.34.1

